drivers/postgresql.c

```diff
@@ -0,0 +1,855 @@
+    exit(1);
+}
+
+extern void
+Mod_db_itr_close(DB_itr_T itr)
+{
+    struct postgresql_handle *dbh = itr->handle->dbh;
+    struct postgresql_itr *dbi = itr->dbi;
+
+    if(dbi) {
+        dbi->curr = NULL;
+        if(dbi->res != NULL)
+            PQclear(dbi->res);
+        free(dbi);
+        itr->dbi = NULL;
+    }
+}
+
+extern int
+Mod_db_itr_next(DB_itr_T itr, struct DB_key *key, struct DB_val *val)
+{
+    struct postgresql_itr *dbi = itr->dbi;
+    PGresult *result = dbi->res;
+
+    itr->current++;
+    if(dbi->res && (itr->current < itr->size)) {
+        populate_key(result, key, 0, itr->current);
+        populate_val(result, val, 4, itr->current);
+        dbi->curr = key;
+        return GREYDB_FOUND;
+    }
+    itr->current--;
+
+    return GREYDB_NOT_FOUND;
+}
+
+extern int
+Mod_db_itr_replace_curr(DB_itr_T itr, struct DB_val *val)
+{
+    struct postgresql_handle *dbh = itr->handle->dbh;
+    struct postgresql_itr *dbi = itr->dbi;
+    struct DB_key *key = dbi->curr;
+
+    return DB_put(itr->handle, key, val);
+}
+
+extern int
+Mod_db_itr_del_curr(DB_itr_T itr)
+{
+    struct postgresql_itr *dbi = itr->dbi;
+
+    if(dbi && dbi->res && dbi->curr)
+        return DB_del(itr->handle, dbi->curr);
+
+    return GREYDB_ERR;
+}
+
+extern int
+Mod_scan_db(DB_handle_T handle, time_t *now, List_T whitelist,
+            List_T whitelist_ipv6, List_T traplist, time_t *white_exp)
+{
+    struct postgresql_handle *dbh = handle->dbh;
+    PGresult *result = NULL;
+    char *sql, *sql_tmpl;
+    int ret = GREYDB_OK;
+    unsigned int size;
+
+    /*
+     * Delete expired entries and whitelist appropriate grey entries,
+     * by un-setting the tuple fields (to, from, helo), but only if there
+     * is not already a conflicting entry with the same IP address
+     * (ie an existing trap entry).
+     */
+    sql_tmpl = "DELETE FROM entries "
+        "WHERE \"expire\" <= EXTRACT(EPOCH FROM now()) "
+        "AND \"greyd_host\"='%s'";
+
+    if(asprintf(&sql, sql_tmpl, dbh->greyd_host) == -1) {
+        i_warning("postgresql asprintf error");
+        goto err;
+    }
+
+    result = PQexec(dbh->db, sql);
+    if(PQresultStatus(result) != PGRES_COMMAND_OK) {
+        i_warning("delete postgresql expired entries: %s",
+                  PQerrorMessage(dbh->db));
+        goto err;
+    }
+    PQclear(result);
+    free(sql);
+    sql = NULL;
+
+    sql_tmpl = "UPDATE entries "
+        "SET \"helo\" = '', \"from\" = '', \"to\" = '', \"expire\" = %lld "
+        "FROM entries e LEFT JOIN entries g "
+        "ON g.\"ip\" = e.\"ip\" AND g.\"to\" = '' AND g.\"from\" = '' "
+        "WHERE e.\"ip\" = entries.\"ip\" AND e.\"helo\" = entries.\"helo\" "
+        "AND e.\"from\" = entries.\"from\" AND e.\"to\" = entries.\"to\" "
+        "AND e.\"from\" <> '' AND e.\"to\" <> '' AND e.\"pcount\" >= 0 "
+        "AND e.\"pass\" <= EXTRACT(EPOCH FROM now()) AND e.\"greyd_host\"='%s' "
+        "AND g.\"ip\" IS NULL";
```

Comment:
See [this](https://www.postgresql.org/docs/9.6/static/queries-with.html) (7.8.2).

The solution might be a statement which will delete the updated rows in a WITH query and upset them back with the appropriate changes. If there is a conflict, then insert the original row which was deleted.